### PR TITLE
Fixes issue with renaming filtered file or directory

### LINKED ISSUE #2475: Renaming filtered file or directory doesn't add it to project tree
This is a bug with the `MsBuildFileSystemWatcher`: When a file or directory is not part of the project tree because it is filtered by an `IMsBuildFileSystemFilter`, and I'm renaming it, then it remains hidden although the filter doesn't apply anymore and it should be visible.

### PR DESCRIPTION
**Pull request by karthiknadig**
Fixes #2475

### CODE REVIEW

File: src/ProjectSystem/Impl/IO/MsBuildFileSystemWatcher.DirectoryRenamed.cs [hunk 1]
```diff
@@ -42,14 +43,37 @@ private class DirectoryRenamed : IFileSystemChange {
                 }
 
                 var oldRelativePath = PathHelper.MakeRelative(_rootDirectory, _oldFullPath);
-                var newRelativePaths = _entries.RenameDirectory(oldRelativePath, newRelativePath, _fileSystem.ToShortRelativePath(_fullPath, _rootDirectory));
+                if (_entries.ContainsDirectoryEntry(oldRelativePath)) {
+                    _entries.RenameDirectory(oldRelativePath, newRelativePath, _fileSystem.ToShortRelativePath(_fullPath, _rootDirectory));
+                } else {
+                    _entries.AddDirectory(newRelativePath, _fileSystem.ToShortRelativePath(_fullPath, _rootDirectory));
+
+                    Queue<string> directories = new Queue<string>();
```
Comment (MikhailArkhipov) [inline src/ProjectSystem/Impl/IO/MsBuildFileSystemWatcher.DirectoryRenamed.cs]: This looks like helper traversal method

**Comment by MikhailArkhipov** [comment]
@AlexanderSher please have a look

File: src/ProjectSystem/Impl/IO/MsBuildFileSystemWatcherEntries.cs [hunk 2]
```diff
@@ -16,6 +16,16 @@ namespace Microsoft.VisualStudio.ProjectSystem.FileSystemMirroring.IO {
     public class MsBuildFileSystemWatcherEntries {
         private readonly Dictionary<string, Entry> _entries = new Dictionary<string, Entry>(StringComparer.OrdinalIgnoreCase);
 
+        public bool ContainsFileEntry(string relativeFilePath) {
+            return _entries.ContainsKey(relativeFilePath);
+        }
+
+        public bool ContainsDirectoryEntry(string relativePath) {
+            relativePath = PathHelper.EnsureTrailingSlash(relativePath);
+            return _entries.Values.Where(v => v.RelativePath.StartsWithIgnoreCase(relativePath) ||
```
Comment (AlexanderSher) [inline src/ProjectSystem/Impl/IO/MsBuildFileSystemWatcherEntries.cs]: `_entries.Values.Any(v => v.RelativePath.StartsWithIgnoreCase(previousRelativePath) ||
                                v.ShortPath.StartsWithIgnoreCase(previousRelativePath))`
And we also can extract this condition into method - it is used in many places

**Comment by AlexanderSher** [comment]
👍

**Pull request merged**